backend/tomato/host.py

```diff
@@ -1029,41 +1079,96 @@ def select(site=None, elementTypes=None, connectionTypes=None, networkKinds=None
 			continue
 		if set(networkKinds) - set(host.getNetworkKinds()):
 			continue
+		if host.detached:
+			continue
 		hosts.append(host)
 	UserError.check(hosts, code=UserError.INVALID_CONFIGURATION, message="No hosts found for requirements")
-	# any host in hosts can handle the request
-	prefs = dict([(h, 0.0) for h in hosts])
-	# STEP 2: calculate preferences based on host load
-	els = 0.0
-	cons = 0.0
-	for h in hosts:
-		prefs[h] -= h.componentErrors * 25  # discourage hosts with previous errors
-		prefs[h] -= h.getLoad() * 100  # up to -100 points for load
-		els += h.elements.count()
-		cons += h.connections.count()
-	avgEls = els / len(hosts)
-	avgCons = cons / len(hosts)
-	for h in hosts:
-		# between -30 and +30 points for element/connection over-/under-population
-		if avgEls:
-			prefs[h] -= max(-20.0, min(10.0 * (h.elements.count() - avgEls) / avgEls, 20.0))
-		if avgCons:
-			prefs[h] -= max(-10.0, min(10.0 * (h.connections.count() - avgCons) / avgCons, 10.0))
-		# STEP 3: calculate preferences based on host location
-	for h in hosts:
-		if h in hostPrefs:
-			prefs[h] += hostPrefs[h]
-		if h.site in sitePrefs:
-			prefs[h] += sitePrefs[h.site]
-	#STEP 4: select the best host
-	hosts.sort(key=lambda h: prefs[h], reverse=True)
-	logging.logMessage("select", category="host", result=hosts[0].name,
+	prefs = dict([(h, getHostValue(h, site, elementTypes, connectionTypes, networkKinds, hostPrefs, sitePrefs)) for h in hosts])
+	return hosts, prefs
+
+def getBestHost(site=None, elementTypes=None, connectionTypes=None,networkKinds=None, hostPrefs=None, sitePrefs=None):
+	hosts_all, prefs = getHostList(site, elementTypes, connectionTypes,networkKinds, hostPrefs, sitePrefs)
+	cand = checkForHostDeactivation()
+	hosts = []
+	for host in hosts_all:
+		if host not in cand:
+			hosts.append(host)
+	if hosts != []:
+		hosts.sort(key=	lambda h: prefs[h], reverse=True)
+	else:
+		hosts = hosts_all.sort(key=	lambda h: prefs[h], reverse=True)
+	return hosts[0], prefs[hosts[0]]
+	
+def checkForHostDeactivation():
+	hosts, prefs = getHostList()
+	candidates = []
+	candidates_prefs = []
+	
+		
+
+	import statistics
+	
+	avg = []
+	for h in getHostList():
+		avg.append(h.getLoad())
+		
+	avg = statistics.mean(avg)
+
+	if avg < AVG_MINIMUM:
+		for host_ in hosts:
+			if host_.detachable:
+				host_elements = HostElement.objects.filter(host = host_)
+				n = 0
+				for el in host_elements:
+					if el.state in ["started"]:
+						n+=1
+				if n == 0:
+					candidates.append(host_)
+					candidates_prefs.append((host_, prefs[host_]))
+		candidates.sort(key=lambda h: prefs[h], reverse=True)
+		
+	return candidates
+	
+def reallocate():
+	
+	import elements
+	#needs to be redefined at a better place
+	THRESHOLD = 20
+	#Walk through all elements and think about reallocating them.
+	for el in  list(elements.getAll()):
+		if el.state in ["started","created"]:
+			continue
+		hostPref, sitePref = el.getLocationPrefs()
```

Comment:
This whole calculation should be done in the element classes in a method ``try_migrate(threshold)``.
The problem here is that not all element types have exactly one main element, so this code does not fit to all types.